lib/kubernetes-deploy/container_logs.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ContainerLogs
+    attr_reader :lines, :container_name
+
+    DEFAULT_LINE_LIMIT = 250
+
+    def initialize(parent_id:, container_name:, logger:)
+      @parent_id = parent_id
+      @container_name = container_name
+      @logger = logger
+      @lines = []
+      @last_printed_index = -1
+    end
+
+    def sync(kubectl)
+      new_logs = fetch_latest(kubectl)
+      return unless new_logs.present?
+      @lines += deduplicate(new_logs)
+    end
+
+    def empty?
+      lines.empty?
+    end
+
+    def print_latest(prefix: false)
+      prefix_str = "[#{container_name}]  " if prefix
+      start_at = @last_printed_index + 1
+
+      lines[start_at..-1].each do |msg|
+        @logger.info "#{prefix_str}#{msg}"
+      end
+
+      @last_printed_index = lines.length - 1
+    end
+
+    def print_all
+      lines.each { |line| @logger.info("\t#{line}") }
+    end
+
+    private
+
+    def fetch_latest(kubectl)
+      cmd = ["logs", @parent_id, "--container=#{container_name}", "--timestamps"]
+      cmd << if @last_timestamp.present?
+        "--since-time=#{rfc3339_timestamp(@last_timestamp)}"
+      else
+        "--tail=#{DEFAULT_LINE_LIMIT}"
+      end
+      out, _err, _st = kubectl.run(*cmd, log_failure: false)
+      out.split("\n")
+    end
+
+    def rfc3339_timestamp(time)
+      time.strftime("%FT%T.%N%:z")
+    end
+
+    def deduplicate(logs)
```

Comment:
Just out of curiosity: Are these logs guaranteed to be returned in order? If so then as soon as we find one line that's not a duplicate, we know that all subsequent lines (from the same API call result) are also not duplicates, right? Seems like knowing this fact would allow us to simplify the code (although I'm not sure yet how).

Reply:
That's a good question. AFAIK k8s is just streaming us the log file written by Docker from the container's stdout/stderr, so it would be super weird for them to get out of order, right? In any case, this code already assumes they're in order, since every iteration overwrites the `@last_timestamp` (assumes ending up with the last timestamp value is correct).

[Logging architecture docs](https://kubernetes.io/docs/concepts/cluster-administration/logging/#logging-at-the-node-level)

Reply:
Did you have an idea of how this could be used to simplify the code? I can certainly see how it could make it more efficient (skip parsing a lot of timestamps), but it seems to me adding awareness of whether the loop had already passed the "outdated" threshold would be more complex, not less. Am I missing something?

Reply:
I added the optimization in my last commit